### PR TITLE
Submitting refined mailroom and mailroom test files for lesson06

### PR DESCRIPTION
**Pull request by ptfbanks**
Test routines prepared.  Each of 5 ran, as far as I could tell, successfuly

### CODE REVIEW

File: students/ptfbanks/Lesson06/pf_mailroom_test.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+amt_1 = 1000.00
+amt_2 = 4000.00
+
+class MyTestCase(unittest.TestCase):
+
+    def test_1(self):  # Tests donors and donations
+        self.assertEqual(['Bob Williams', 'Tom Haskell', 'Earl Jackson',
+                          'Les Thomas', 'James Black', 'Ed Jones'], list(d_dict.keys()))
+        self.assertTrue(d_dict['Bob Williams'] == [5580.00, 3245.50, 1000])
+        print("End Test 1---------------------------------------------")
```
Comment (natasha-aleksandrova) [inline students/ptfbanks/Lesson06/pf_mailroom_test.py]: this test is necessary as it is testing the python dictionary

File: students/ptfbanks/Lesson06/pf_mailroom_test.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    def test_2(self):  # Tests add donors and donations             
+        d_dict[n_name]=[float(amt_1)]
+#        print(d_dict)  #Diagnostic insert
+        self.assertTrue(d_dict['New Mann'] == [1000.0])
+        print("End Test 2---------------------------------------------")
+       
+    def test_3(self):  # Tests add donation            
+        d_dict[n_name].append(float(amt_2))
+#        print(d_dict)    #Diagnostic insert
+        self.assertTrue(d_dict['New Mann'] == [1000.0, 4000.0])
```
Comment (natasha-aleksandrova) [inline students/ptfbanks/Lesson06/pf_mailroom_test.py]: same comment here: this is testing the `.append` method which is a python core functionality

File: students/ptfbanks/Lesson06/pf_mailroom_test.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                  "It will be put to very good use.\n\n\t\t Sincereley,\n"
+                  "\t\t   -The Team")
+#        print(n_name, amt, "\n",output)      #Diagnostic insert
+#        print (letter.format(n_name, amt))   #Diagnostic insert
+        self.assertTrue(letter.format(n_name, amt) == output)
+        print("End Test 4---------------------------------------------")        
+        
+    def test_5(self): # Confirm .txt file is generated
+        send_where (n_name, 'y', "")
+        assert os.path.isfile('New Mann.txt')
```
Comment (natasha-aleksandrova) [inline students/ptfbanks/Lesson06/pf_mailroom_test.py]: good test 👍

File: students/ptfbanks/Lesson06/pf_mailroom_test.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        
+    def test_4(self):# Confirm letter is properly prepared
+        amt = sum(d_dict[n_name])
+        output = ("Dear New Mann,\n\t Thank you for your very kind donation(s) totaling $ 5000.0.\n"
+                  "It will be put to very good use.\n\n\t\t Sincereley,\n"
+                  "\t\t   -The Team")
+#        print(n_name, amt, "\n",output)      #Diagnostic insert
+#        print (letter.format(n_name, amt))   #Diagnostic insert
+        self.assertTrue(letter.format(n_name, amt) == output)
+        print("End Test 4---------------------------------------------")
```
Comment (natasha-aleksandrova) [inline students/ptfbanks/Lesson06/pf_mailroom_test.py]: this isn't testing your mailroom code; this as well only tests core python language which don't need to be tested.

File: students/ptfbanks/Lesson06/pf_mailroom_test.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        amt = sum(d_dict[n_name])
+        output = ("Dear New Mann,\n\t Thank you for your very kind donation(s) totaling $ 5000.0.\n"
+                  "It will be put to very good use.\n\n\t\t Sincereley,\n"
+                  "\t\t   -The Team")
+#        print(n_name, amt, "\n",output)      #Diagnostic insert
+#        print (letter.format(n_name, amt))   #Diagnostic insert
+        self.assertTrue(letter.format(n_name, amt) == output)
+        print("End Test 4---------------------------------------------")        
+        
+    def test_5(self): # Confirm .txt file is generated
```
Comment (natasha-aleksandrova) [inline students/ptfbanks/Lesson06/pf_mailroom_test.py]: it is best to name your test functions with meaningful names like: `test_send_where` or `test_send_where_creates_file` for a more verbose option.

**Comment by ptfbanks** [comment]
Thank you.   suppose it will take a bit of experience and exposure to sense
the difference of testing core functionality from the testing the
application of functionality.  I may have sensed that somewhat so I laced
through the test steps the application of a test subject, "Neut Mann."
This was intended represent how each element of the system was exercised,
although not directly.

I did have to refer to other student's examples I suspect  may have
perpetuated flaws in their solutions.  I am not sure how tell where to find
proper examples to study.

I understand that my limited function mailroom program could/should be
refined to reflect more appropriate coding disciplines and practices. That
will com in time- I hope

**Comment by natasha-aleksandrova** [comment]
@ptfbanks this assignment is one of the more challenging ones for most students. I've been programming for some time and still sometimes struggle with unit-testing (in reference to large scale applications). It's one thing to test neat little function that take some input and then return something back - but it's another level to test a program like mailroom.

The big take away from this assignment is realizing that it requires a code refactor in order to make it unit-testable. In addition to that, many parts of the code that deal with `input` functions would not be testable in this course as it requires advanced concepts that use `mock` library; it's part of the next course.

I would recommend reading this over:
https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/mailroom-part4.html

This is a new content that was recently added and has some hints.

And of course don't hesitate to email us if need help on assignments.

**Pull request merged**